File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/ClassTypeTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;
import com.loudsight.useful.helper.ClassHelper;

import java.util.Iterator;
import java.util.List;

public final class ClassTypeTransform extends EntityTransform<Class<?>> {

private static final class ClassTypeTransformHolder {
    private static final ClassTypeTransform INSTANCE = new ClassTypeTransform();
}
    // global access point
    public static ClassTypeTransform getInstance() {
        return ClassTypeTransformHolder.INSTANCE;
    }
    private ClassTypeTransform() {
        super(EntityType.CLASS, ClassHelper.<Class<Class<?>>>uncheckedCast(Class.class));
    }

    @Override public void serializeEntity( Class<?> entity, List<Byte> bytes) {
        bytes.add(EntityType.CLASS.getCode());
        writeStr(entity.getName(), bytes);
    }

    @Override public Class<?>  deserializeEntity(Iterator<Byte> bytes) {
        var className = readStr(bytes);

        try {
            return Class.forName(className);
        } catch (ClassNotFoundException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/ClassTypeTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType
import kotlin.reflect.KClass

object ClassTypeTransform : EntityTransform<Class<*>>(EntityType.CLASS, Class::class) {

    override fun serializeEntity(entity: Class<*>, bytes: MutableList<Byte>) {
        bytes.add(EntityType.CLASS.code)
        writeStr(entity.name, bytes)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Class<*> {
        val className = readStr(bytes)

        return Class.forName(className)
    }

    override fun canTransform(entity: Any): Boolean {
        return entity is Class<*>
    }
}